History (Undo, Redo, Reset)

martanman opened the pull request.

martanman committed:
added history for selectedclasses

martanman committed:
made it set courses as well with un/redo

martanman committed:
proper rerender when courses change (un/redo)

martanman committed:
code clean up, disable un/redo if end reached

martanman committed:
map ctrl z and ctrl y to un/redo, put reset button

martanman committed:
updated type from any to KeyboardEvent

munjoonteo reviewed (changes requested):
1. Clicking the reset button adds an action every time i.e. click reset 5 times, have to undo 5 times. Would it be possible to make it so that no matter how many times you click reset you only have to undo once?
2.  When you drag a class around, every time the class enters a valid dropzone, an action is added to the list. So when you undo you kinda retrace the direction in which you dragged the class. Can an action be added to the list only after the class is dropped into a dropzone?

martanman committed:
waits for drag to end before updating

- starting a drag sets isDrag to true (by the droppedClass component that initiated it), and in this state the history actions won't register any changes to selected classes until isDrag is falsified (triggered by mouseup event).
- before adding a new action to the history actions, it will check that nextAction !== prevAction
- wrapped un/redo buttons in a span because mui got angry at me for it (and it resulted in a semi-bug where if you click undo and redo a lot the tooltip may decide not to disappear even when mouse moves off)

martanman committed:
Merge branch 'dev' into history

martanman committed:
made identicality check skip initial loading

Resulted in a bug where the initial state wasn't registered to history if no classes selected initially.

martanman committed:
moved next to autotimetabling

added ctrl z / y tooltip

martanman committed:
chucked history actions next to autotimetabler

munjoonteo committed:
minor fixes
